hypervisor/arch/x86/virq.c

```diff
@@ -261,6 +261,10 @@ static void vcpu_inject_exception(struct vcpu *vcpu, uint32_t vector)
 			(exception_type[vector] << 8U) | (vector & 0xFFU));
 
 	vcpu->arch_vcpu.exception_info.exception = VECTOR_INVALID;
+
+	/* exception with class fault returns to the faulting instruction */
+	if (FAULT_MASK & (1U << vector))
+		vcpu_retain_rip(vcpu);
```

Comment:
According to the SDM 6.15, there is difference that push EIP for fault and trap.
For example:
Divide Error Exception (DE fault): 
Saved contents of CS and EIP point to the instruction that generate the exeptino.

But for Breakpoint Exception (BP trap):
Saved contents of CS and EIP registers point to the instruction following the INT 3 instruction.

My understanding is: we should keep RIP unchanged for Fault exception and change RIP to next instruction for trap exception. So here we only retain RIP for fault exceptions.

Reply:
@fyin1 we must already retain rip for PF, otherwise it won't work if there is a PF when guest access MMIO.
does ACRN inject TRAP to guest?

Reply:
I didn't find where we retain the rip for PF. Yes. We have code to inject TRAP to guest. It's in vcpu_inject_hi_exception. But it may not real case to inject TRAP.

But re-think the exception, I suppose we should retain RIP for all exception. It's VMX duty to push the EIP of instruction following the faulting instruction for trap. What we inject is showing the faulting instruction EIP. I will update the patch to remove the condition.